src/NuGet.Core/NuGet.Packaging/ContentModel/ContentItemCollection.cs

```diff
@@ -11,6 +12,7 @@ namespace NuGet.ContentModel
     public class ContentItemCollection
     {
         private List<Asset> _assets;
+        private ConcurrentDictionary<string, string> _assemblyRelatedExtensions;
```

Comment:
Do you expect this item going to be mutated from different threads?
Please note it only matters for Add/Update/Delete operation not for Get value from it.

Reply:
Yes. I first used `Dictionary<string, string>` instead of `ConcurrentDictionary<string, string>`. Then the patched dotnet SDK constantly failed the restore for NuGet.sln as following:
```
C:\PatchedSDK3\sdk\7.0.100-preview.4.22252.9\NuGet.targets(131,5): error : An item with the same key has already been added. Key: ref/netstandard1.3/System.IO.FileSystem.DriveInfo. [C:\Users\henli\repos\NuGet.Client\NuGet.sln]

C:\PatchedSDK3\sdk\7.0.100-preview.4.22252.9\NuGet.targets(131,5): error : Operations that change non-concurrent collections must have exclusive access. A concurrent update was performed on this collection and corrupted its state. The collection's state is no longer correct. [C:\Users\henli\repos\NuGet.Client\NuGet.sln]
```

Reply:
For a solution with multiple projects, running restore on the solution will trigger multiple threads(per project), each thread will call [BuildAssetsFile ](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs#L285)-> CreateLockFile -> CreateLockFileTargetLibrary.
Add/Get the `_assemblyRelatedExtensions` is in [AddAssets](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/LockFileUtils.cs#L106), (GetLockFileItems -> FindBestItemGroup ->PopulateItemGroups ->FindItemsImplementation ->GetRelatedFileExtensionProperty).
So if there are multiple projects referencing the same package, there might be multiple threads trying to add into _assemblyRelatedExtensions with the same key(assembly path).